src/errors.rs

```diff
@@ -47,6 +50,168 @@ impl Display for ErrorType {
     }
 }
 
+/// Custom result class, that supports extended errors.
+/// Returning Error variants will result in throwing errors in JS.
+///
+/// You can either use Converted error, that supports some extra fields,
+/// or you can add a new error type. Note that `with_custom_error!` macro
+/// supports only ConvertedError conversion. Usage with different error requires manual handling.
+/// This is important, because we cannot implement ToNapiValue for Result<T, E> directly,
+/// And we cannot implement FromResidual for this class, as this is unstable feature.
+///
+/// When converting this class to JS value through ToNapiValue
+/// cases will be handled in the following way:
+/// - Ok(T): T::to_napi_value will be returned
+/// - Error(ConvertedError): We will create napi::Error based on ConvertedError ToNapiValue trait and return Err(Error)
+/// - NapiError(Error): Err(Error) will be returned
+pub enum JsResult<T> {
+    Ok(T),
+    Error(ConvertedError),
+    NapiError(napi::Error),
+}
+
+impl<T, E> From<Result<T, E>> for JsResult<T>
+where
+    E: Error,
+{
+    fn from(value: Result<T, E>) -> Self {
+        match value {
+            Ok(v) => JsResult::Ok(v),
+            Err(e) => JsResult::Error(e.into()),
+        }
+    }
+}
+
+pub(crate) trait ToConvertedError<Final> {
+    fn cnv(self) -> Final;
+}
```

Comment:
:wrench: Customary name for such conversion would be `to_converted_error`.

:wrench: Also, `to_*` methods take `&self`, while `into_*` take `self`. Once again please take a look at https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv.

:thought_balloon: Shouldn't `Final` be an associated type instead of generic type parameter?